src/lib/transaction_snark_scan_state/transaction_snark_scan_state.ml

```diff
@@ -345,69 +365,116 @@ end = struct
         ; fee_excess
         ; proof_type= `Merge }
 
-  let fill_in_transaction_snark_work t (works : Transaction_snark_work.t list)
-      : Ledger_proof.t option Or_error.t =
-    let open Or_error.Let_syntax in
-    let%bind next_jobs =
-      Parallel_scan.next_k_jobs ~state:t ~k:(total_proofs works)
-    in
-    let%bind scanable_work_list =
-      map2_or_error next_jobs
-        (List.concat_map works
-           ~f:(fun {Transaction_snark_work.fee; proofs; prover} ->
-             List.map proofs ~f:(fun proof -> (fee, proof, prover)) ))
-        ~f:completed_work_to_scanable_work
-    in
-    let%map result =
-      Parallel_scan.fill_in_completed_jobs ~state:t
-        ~completed_jobs:scanable_work_list
-    in
-    Option.map result ~f:fst
-
-  let capacity t = Parallel_scan.parallelism ~state:t
+  let capacity t = Parallel_scan.parallelism ~state:t.tree
 
   let create ~transaction_capacity_log_2 =
-    Parallel_scan.start ~parallelism_log_2:(transaction_capacity_log_2 + 2)
+    (* Transaction capacity log_2 is 1/2^work_delay_factor the capacity for work parallelism *)
+    let k = max Config.work_delay_factor 2 in
+    { tree=
+        Parallel_scan.start ~parallelism_log_2:(transaction_capacity_log_2 + k)
+    ; job_count= 0 }
 
   let empty () =
     let open Config in
     create ~transaction_capacity_log_2
 
-  let enqueue_transactions t transactions =
-    Parallel_scan.enqueue_data ~state:t ~data:transactions
+  let fill_work_and_enqueue_transactions t transactions work =
+    let open Or_error.Let_syntax in
+    let enqueue_transactions t transactions =
+      Parallel_scan.enqueue_data ~state:t ~data:transactions
+    in
+    let fill_in_transaction_snark_work t
+        (works : Transaction_snark_work.t list) :
+        Ledger_proof.t option Or_error.t =
+      let%bind next_jobs =
+        Parallel_scan.next_k_jobs ~state:t ~k:(total_proofs works)
+      in
+      let%bind scanable_work_list =
+        map2_or_error next_jobs
+          (List.concat_map works
+             ~f:(fun {Transaction_snark_work.fee; proofs; prover} ->
+               List.map proofs ~f:(fun proof -> (fee, proof, prover)) ))
+          ~f:completed_work_to_scanable_work
+      in
+      let%map result =
+        Parallel_scan.fill_in_completed_jobs ~state:t
+          ~completed_jobs:scanable_work_list
+      in
+      Option.map result ~f:fst
+    in
+    let work_count =
+      List.sum
+        (module Int)
+        work
+        ~f:(fun (w : Transaction_snark_work.t) -> List.length w.proofs)
+    in
+    let%bind () = Parallel_scan.update_curr_job_seq_no t.tree in
+    let%bind proof_opt = fill_in_transaction_snark_work t.tree work in
+    let%bind () = enqueue_transactions t.tree transactions in
+    (*important: Everytime a proof is emitted, reduce the job count by 1 because you only had to do (2^x - 1 extra jobs). This is important because otherwise the job count would never become zero*)
+    let adjust_job_count =
```

Comment:
Created an issue (#1531) to later decide on what strategy the protocol should be taking here